### PR TITLE
Java sample program

### PR DESCRIPTION
**Pull request by KeithRieck**
I've added a sample application for running GRIP pipelines in Java.  I have tested this on Windows, Mac, Ubuntu, and Raspbian.

### CODE REVIEW

File: java/src/main/java/edu/wpi/first/wpilibj/vision/VideoViewer.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+		Image loadedImage = toBufferedImage(image);
+		imageView.setIcon(new ImageIcon(loadedImage));
+		frame.pack();
+
+	}
+
+	private void setSystemLookAndFeel() {
+		try {
+			UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
+		} catch (ClassNotFoundException e) {
```
Comment (SamCarlberg) [inline java/src/main/java/edu/wpi/first/wpilibj/vision/VideoViewer.java]: Use a multicatch

File: java/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+* ANT_HOME = the directory in which Apache Ant is installed
+* OPENCV_HOME = the directory in which all of OpenCV is installed
+* OPENCV_LIB = the directory containing all native libraries
+* OPENCV_JAR = the path to the JAR file containing the java interface to OpenCV (typically named something like "opencv-320.jar" )
+
+Execute the code by changing to the directory for GRIP-code-generation/java and then executing "ant run".   Or, can can run the application by executing:
+
+`java -Djava.library.path=${OPENCV_LIB} -jar build/grip_sample1.jar`
+
+## Installing OpenCV on Windows
```
Comment (SamCarlberg) [inline java/README.md:27]: Could mention the opencv-installer program we've written that downloads and installs WPI's prebuilt bindings. No need to compile OpenCV yourself (for the most part -- it only does x86_64 windows, linux, and OS X, with support for generic linux armhf and rasbian coming soon)

File: java/src/main/java/grip/sample1/FindRedAreasApp.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import edu.wpi.first.wpilibj.vision.GripRunner.Listener;
+import edu.wpi.first.wpilibj.vision.VideoViewer;
+
+public class FindRedAreasApp {
+	
+	static final int IMG_WIDTH = 320;
+	static final int IMG_HEIGHT = 240;
+
+	final VideoViewer window;
+	final Listener<? super FindRedAreas> listener;
```
Comment (SamCarlberg) [inline java/src/main/java/grip/sample1/FindRedAreasApp.java]: This should just be

```java
final Listener<FindRedAreas> listener;
```

File: java/src/main/java/grip/sample1/FindRedAreasApp.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	final GripRunner<FindRedAreas> gripRunner;
+	
+	public FindRedAreasApp() {
+		this.window = makeWindow("GRIP", IMG_WIDTH, IMG_HEIGHT);
+		this.listener = (this.window!=null) ? (processor -> { window.imshow(processor.rgbThresholdOutput());}) : null;
+		this.gripRunner = new GripRunner<FindRedAreas>(
```
Comment (SamCarlberg) [inline java/src/main/java/grip/sample1/FindRedAreasApp.java]: `<FindRedAreas>` is redundant, just use the diamond operator `<>`

**Review by SamCarlberg — CHANGES_REQUESTED** [review]
Thanks for doing this! There's a few small things I'd like addressed, but on the whole this is very good

File: java/src/main/java/edu/wpi/first/wpilibj/vision/GripRunner.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+	/**
+	 * A convenience method that calls {@link #runOnce()} in an infinite loop.
+	 * This must be run in a dedicated thread, and cannot be used in the main
+	 * robot thread because it will freeze the robot program.
+	 * 
+	 * @see VisionThread
+	 */
+	public void runForever() {
+		while (true) {
```
Comment (JLLeitschuh) [inline java/src/main/java/edu/wpi/first/wpilibj/vision/GripRunner.java]: This should check if the thread's interrupted flag has been set.
```java
while(!Thread.currentThread().isInterrupted())
```

**Comment by JLLeitschuh** [comment]
This overall looks really good. Thank you for your contribution.
If you could clean up the things that I and @SamCarlberg have requested that would be awesome!

**Comment by KeithRieck** [comment]
OK, I've pushed fixes.

**Comment by KeithRieck** [comment]
OK, I think I've got everything fixed.
I'm not sure if I need to resubmit the pull request though. ---- Keith Rieck rieck_55113@yahoo.com

**Review by SamCarlberg — APPROVED** [review]

**Pull request merged**